src/utils/RequestQueue.ts

```diff
@@ -0,0 +1,271 @@
+  private activeRequests: Set<string>;
+
+  /** Used to convert keys into a string identifier. Can be overridden in constructor. */
+  private keyStringifyFn: (key: K) => string;
+
+  /**
+   * Creates a new RequestQueue.
+   * @param keyStringifyFn Function used to turn the key type into a string identifier. By default, uses JSON.stringify.
+   * @param maxActiveRequests The maximum number of requests that will be handled concurrently. This is 10 by default.
+   */
+  constructor(keyStringifyFn = (key: K) => JSON.stringify(key), maxActiveRequests = 10) {
+    this.allRequests = new Map();
+    this.activeRequests = new Set();
+    this.queue = [];
+    this.maxActiveRequests = maxActiveRequests;
+    this.keyStringifyFn = keyStringifyFn;
+  }
+
+  /**
+   * Stores request metadata to the internal map of all pending requests.
+   * @param keyString string identifier of the request.
+   * @param requestAction callable function action of the request.
+   * @returns the promise of the request, to be resolved later.
+   */
+  private registerRequest<T>(keyString: string, requestAction: () => Promise<T>): Promise<T> {
+    // Create a new promise and store the resolve and reject callbacks for later.
+    // This lets us perform the actual action at a later point, when the request is at the
+    // front of the processing queue.
+    let promiseResolve, promiseReject;
+    const promise = new Promise<T>((resolve, reject) => {
+      promiseResolve = resolve;
+      promiseReject = reject;
+    });
+    // Store the request data.
+    this.allRequests.set(keyString, {
+      key: keyString,
+      action: requestAction,
+      resolve: promiseResolve,
+      reject: promiseReject,
+      promise,
+    });
+    return promise;
+  }
+
+  /**
+   * Adds a request with a unique key to the queue, if it doesn't already exist.
+   * @param key The key used to track the request. Keys will be compared using the `keyStringifyFn` function provided
+   *  in the constructor.
+   * @param requestAction Function that will be called to complete the request. The function
+   *  will be run only once per unique key while the request exists, and may be deferred by the
+   *  queue at any time.
+   *
+   * NOTE: Cancelling a request while the action is running WILL NOT stop the action. If this behavior is desired,
+   * actions must be responsible for checking the RequestQueue, determining if the request is still valid (e.g.
+   * using `.hasRequest()`), and stopping or returning early.
+   *
+   * @returns A promise that will resolve on completion of the request, or reject if the request is cancelled.
+   *  If multiple requests are issued with the same key, a promise for the first request will be returned
+   *  until the request is resolved or cancelled.
+   *  Note that the return type of the promise will match that of the first request's instance.
+   */
+  public addRequest<T>(key: K, requestAction: () => Promise<T>): Promise<unknown> {
```

Comment:
could this take an optional delay parameter as its third arg?

Reply:
Added, refactored some other code to be cleaner too 👍